WINC-1174: WinC Disconnected Support (#1531)
openshift/enhancements: enhancements/windows-containers/disconnected-support.md

---
title: windows-node-egress-proxy
authors:
  - "@saifshaikh48"
reviewers:
  - "@openshift/openshift-team-windows-containers"
  - "@openshift/openshift-team-pod-autoscaling, for insight on MCO's registry configuration on Linux nodes"
approvers:
  - "@aravindhp"
api-approvers:
  - None
creation-date: 2023-12-11
last-updated: 2023-12-11
tracking-link:
  - "https://issues.redhat.com/browse/OCPSTRAT-619"
  - "https://issues.redhat.com/browse/WINC-936"
see-also:
  - "https://github.com/openshift/enhancements/api-review/add-new-CRD-ImageDigestMirrorSet-and-ImageTagMirrorSet-to-config.openshift.io.md"

>   - Disconnected investigation doc: "https://docs.google.com/document/d/1FYpaEA8gN4Mk4xTRITRLq04gZ_GA_ZN0Xbxii9BkS_E/edit?usp=sharing"

Reviewer: Agreed

  - https://cloud.redhat.com/blog/is-your-operator-air-gap-friendly
---

# Windows Containers Support in Disconnected Environments

## Release Sign-off Checklist

- [x] Enhancement is `implementable`
- [x] Design details are appropriately documented from clear requirements
- [x] Test plan is defined
- [x] Operational readiness criteria is defined
- [x] Graduation criteria for dev preview, tech preview, GA

> - [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

Author: I think the intent is to mark it as [x] when official docs are published

environments where Windows nodes need to pull images from air-gapped, on-prem registries for security reasons.
The added benefit here is that the changes proposed in this enhancement are not restricted to disconnected environments;
users in __connected__ clusters can take advantage of this mirror registry support.

### Goals

* Create an automated mechanism for WMCO to consume global mirror registry settings from existing platform resources,
  such that it enables Windows nodes to pull images in disconnected environments
* Configure the mirror registry settings in WMCO-managed components on Windows nodes, namely the containerd runtime
* React to changes to the mirror registry settings during WMCO runtime
* Maintain normal functionality in connected clusters and clusters without any cluster-wide mirror registry configuration

### Non-Goals

* First-class support of [all OpenShift image registry setting configurations](https://docs.openshift.com/container-platform/4.14/openshift_images/image-configuration.html#images-configuration-file_image-configuration), including the `Image` resource

> * Supporting clusters that use only `ImageContentSourcePolicy` resources. These were [deprecated in 4.13](https://docs.openshift.com/container-platform/4.14/release_notes/ocp-4-14-release-notes.html#ocp-4-14-deprecated-removed-features) 

Author: MCO still consumes ImageContentSourcePolicy despite it being deprecated, so I just wanted to be clear we won't

> * Supporting clusters that use only `ImageContentSourcePolicy` resources. These were [deprecated in 4.13](https://docs.openshift.com/container-platform/4.14/release_notes/ocp-4-14-release-notes.html#ocp-4-14-deprecated-removed-features) 

Reviewer 2: :+1: Let's stick with the new CRDs here.

> and users [should migrate to IDMSes](https://docs.openshift.com/container-platform/4.14/openshift_images/image-configuration.html#images-configuration-registry-mirror-convert_image-configuration) instead.

Author: sure I made it more explicit in the Goals

> * Using [`ImageContentPolicy` resources](https://docs.openshift.com/container-platform/4.14/rest_api/config_apis/imagecontentpolicy-config-openshift-io-v1.html)

Author: Out of scope for the disconnected MVP

> * Using [`ImageContentPolicy` resources](https://docs.openshift.com/container-platform/4.14/rest_api/config_apis/imagecontentpolicy-config-openshift-io-v1.html)

Reviewer 3: I understand this is out of scope but it will help to state why. This is going to be an obvious follow up question from a customer and product management and it will be good to have it documented.

apiVersion: config.openshift.io/v1
kind: ImageTagMirrorSet
metadata:
  name: tag-mirror
spec:
  imageTagMirrors:
  - mirrors:
    - docker.io
    source: docker-mirror.internal
    mirrorSourcePolicy: AllowContactingSource
```

would result in WMCO creating 3 files to place on each Windows instance.
```powershell
$ tree $config_path
C:/k/containerd/registries/
|── registry.access.redhat.com
|   └── hosts.toml
|── mirror.example.com
|   └── hosts.toml
└── docker.io
    └── hosts.toml

$ cat "$config_path"/registry.access.redhat.com/host.toml
server = "https://registry.access.redhat.com" # default fallback server since "AllowContactingSource" mirrorSourcePolicy is set

[host."https://example.io/example/ubi-minimal"]
 capabilities = ["pull"]

[host."https://example.com/example2/ubi-minimal"] # secondary mirror
 capabilities = ["pull"]

$ cat "$config_path"/registry.redhat.io/host.toml
# "server" omitted since "NeverContactSource" mirrorSourcePolicy is set

[host."https://mirror.example.com"]
 capabilities = ["pull"]

$ cat "$config_path"/docker.io/host.toml
server = "https://docker.io"

[host."https://docker-mirror.internal"]
 capabilities = ["pull", "resolve"] # resolve tags
```

> As on the Linux side, it is upon the user to ensure there is no clash between mirrors across their IDMS and ITMS resources.

Author: I lean towards no as this should probably be done by something at the cluster level, not WMCO

---

### Test Plan & Infrastructure Needed

In addition to unit testing individual WMCO/WICD packages and controllers, an e2e job will be added to the release repo
for WMCO's master/release-4.15 branches. A new CI workflow will be created leveraging 
[existing step-registry steps](https://github.com/openshift/release/tree/master/ci-operator/step-registry/aws/provision/vpc/disconnected),

> which creates a disconnected cluster in AWS or vSphere with hybrid-overlay networking. 

Reviewer 4: >  AWS or vSphere

pros and cons?

> which creates a disconnected cluster in AWS or vSphere with hybrid-overlay networking. 

Reviewer 4: > hybrid-overlay networking

is a WMCO requirement, not sure about the value added by mentioning it here.

Author: value here is that the existing workflows in the step registry don't bring up clusters with OVN hybrid out of the box. We'll have to add this step into our job's steps

> We'll also need to pre-load the images used in our e2e suite into the cluster's internal registry as part of job setup.

Reviewer 4: Is this possible?

Author: I have not tried but we should be able to use `oc mirror` or `skopeo` commands to copy over images into the internal registry at install time

Reviewer 3: Are you sure the internal registry supports hosting Window images? This needs to be clarified in the design section.

Author: In the past the only issue was the foreign layer links, but now those are removed and images carry the full load when pushed to a registry. I asked the image registry folks to make sure, waiting on a response

Author: Confirmed with the image registry folks (and a customer) that OCI compliant Windows images can be hosted fine by the internal registry, existing OCP docs should suffice on how to do this

This workflow will be used to run the existing WMCO e2e test suite to ensure that WMCO runs as expected in disconnected 
environments with no regression. We will add a few test cases to explicitly check the state of registry settings on Windows
nodes.

> QE should cover disconnected envrionments on all platforms when validating this feature.

Reviewer 4: ```suggestion
QE should cover disconnected environments on all supported platforms when validating this feature.
```

### User Stories

<!--
Story 1: set proper config_path in containerd_conf.toml
Story 2: generate hosts.toml(s) at config_path from cluster's IDMS/ITMS resources

> -->

Reviewer 4: Will CI and e2e integration covered in separate stories?

User stories can also be found within the JIRA epic: [WINC-936](https://issues.redhat.com/browse/WINC-936)

### Workflow Description and Variations

**cluster creator** is a human user responsible for deploying a cluster.
**cluster administrator** is a human user responsible for managing cluster settings including network egress policies.

There are 3 different workflows that affect the disconnected use case.
1. A cluster creator brings up a new disconnected cluster and configures mirror registry settings at install time
2. A cluster administrator introduces new mirror registry settings or changes exisiting ones during runtime
   This would occur through the creation, deletion, or update IDMS/ITMS resources.

> 3. A cluster administrator transitions a connected cluster *with existing mirror registry settings* to a disconnected env

Reviewer 4: ```suggestion
3. A cluster administrator transitions a connected cluster *with existing mirror registry settings* to a disconnected cluster
```

> 4. A cluster administrator transitions a connected cluster *without existing mirror registry settings* to a disconnected env

Reviewer 4: What about transitions from disconnected to connected?

Author: should behave the same way as connected to disconnected

In cases 1, 2 & 3, Windows nodes can be joined to the cluster after altering registry settings,
which would result in WMCO applying mirror settings during initial node configuration. 

In case 2, for any existing Windows nodes, WMCO will react by updating each instance's state to consume the new registry config.

In case 3, existing Windows nodes will not be re-configured. Existing registry settings will stay applied.

In case 4, WMCO will take no action as there are no settings to configure. Windows nodes will not be able to pull images
until mirror registry settings are configured by the administrator.

### Risks and Mitigations

> Containerd does not support all the fields available in IDMS/ITMS resources.

Reviewer 4: Would this result in WMCO unable to generate/transfer the registry configuration to the Windows instances?

> If node cleanup fails, instances may retain their containerd configuration. This could be a concern for BYOH instances.

Reviewer 4: Mitigation? Documentation for manual cleanup, maybe?

but linking to existing docs should be enough for overarching info such as mirroring images to disconnected environments,
creating image set configurations (IDMS/ITMS resouces), and authenticating with private registries
(e.g. [using image pull secrets](https://docs.openshift.com/container-platform/4.14/openshift_images/managing_images/using-image-pull-secrets.html)).

### Graduation Criteria

#### Dev Preview -> Tech Preview

N/A

#### Tech Preview -> GA

See Release Plan above.

#### Removing a deprecated feature

N/A, as this is a new feature that does not supersede an existing one.

### Upgrade / Downgrade Strategy

The relevant upgrade path is from WMCO 9.y.z in OCP 4.14 to WMCO 10.15.0 in OCP 4.15. There will be no changes to the
current WMCO upgrade strategy. Once customers are on WMCO 10.15.0, they can [convert their connected cluster to a
disconnected one](https://docs.openshift.com/container-platform/4.14/post_installation_configuration/connected-to-disconnected.html#connected-to-disconnected-config-registry_connected-to-disconnected)
and the Windows nodes will be automatically updated by the operator to use the mirror settings to pull required images.

> When deconfiguring Windows instances, mirror settings will be cleared from the node. This involves undoing some node

Reviewer 4: ```suggestion
When deconfiguring Windows nodes, mirror settings will be cleared from the Windows instances. This involves undoing some node
```

config steps, which will be taken care of as part of removing the `containerd` Windows service.

> This scenario will occur when upgrading both BYOH and Machine-backed Windows nodes.

Reviewer 4: ```suggestion
This scenario will occur when upgrading both BYOH and Machine-backed Windows instances.
```

Downgrades are generally [not supported by OLM](https://github.com/operator-framework/operator-lifecycle-manager/issues/1177),
which manages WMCO. In case of breaking changes, please see the
[WMCO Upgrades](https://github.com/openshift/enhancements/blob/master/enhancements/windows-containers/windows-machine-config-operator-upgrades.md#risks-and-mitigations)
enhancement document for guidance.

### Version Skew Strategy

N/A. There will be no version skew since this work is all within 1 product sub-component (WMCO). 
The official Red Hat operator in OCP 4.15, WMCO version 10.15.0, will support Windows containers in disconnected
environments. Releases 9.y.z and below will not have support disconnected support.

## Implementation History

The implementation history can be tracked by following the associated work items in Jira and source code improvements in
the WMCO Github repo.

## Alternatives & Justification
 
For disconnected support, we could make the kubelet pause image location configurable. This is a nice to have, but may 
be out of scope of disconnected support as it could require a larger undertaking to make the containerd config file

> configurable.  A valid workaround exists, which is instructing users to set a mirror (through IDMS/ITMSes) for the MSFT

Reviewer 4: > valid workaround exists

missing ref

external registry where `mcr.microsoft.com/oss/kubernetes/pause:3.9` lives.

There are a few alternatives that are no longer viable.
Pre-pulling images, where user creates a Windows image with the pause image and other required images baked in, is not
possible anymore as we install and manage the container runtime.

> Also, users cannot easily configure containerd's registries manually since WMCO will overwrite changes to the containerd

Reviewer 4: ```suggestion
Also, users cannot configure containerd's registries manually since WMCO will overwrite changes to the containerd
```

config file if there is some unrelated event that triggers node reconciliation.

## Open Questions

> * Enabling pulling from private registries that require authentication credentials or additional certificates?

Reviewer 4: Will private registries requiring authentication be a release blocker feature?

Author: maybe, I added some info on this below

  * Would additional auth like image pull secrets, creds, CAs, etc be usable out of the box through the Internal Registry Operator?
  “When referencing the Windows image in deployments or pods, use the repository and version tag that you specified while tagging the image
  OpenShift will handle the appropriate image pulling and deployment for Windows-based workloads (e.g. image pull secrets). 
  Image Registry Operator should be sufficient to handle image pull config just as it does for Linux images.”

> * Would it work "out-of-the-box" today if images were pre-loaded into the OpenShift internal registry,

Reviewer 4: Do pre-loaded images need specific tags so that WMCO and workloads can pull them?

Author: yes, the information on this is below

  then containerd was configured to use the internal registry as a [default mirror for all registries](https://github.com/containerd/containerd/blob/main/docs/hosts.md#setup-default-mirror-for-all-registries)? 

Minimal work required for us:
We'd have to set a `config_path` in the base containerd_conf.toml file to a valid value, as recommended in their docs.
It is currently set to `""` -- containerd log output:
```
Registry:{ConfigPath: Mirrors:map[] Configs:map[] Auths:map[] Headers:map[]}
```
This serves as the location of the _default/hosts.toml file, which, if the user set manually today, 
would be stomped back by WMCO if there were any unrelated events that triggers node reconciliation.

Bit of manual work on the user:

> 1. Put images in private registry (air-gapped, on-prem or openshift internal registry) before disconnecting from internet: kubernetes pause image (tagged) + any images required for user workloads

Reviewer 4: ```suggestion
1. Pull and tag images in private registry (air-gapped, on-prem or openshift internal registry) before disconnecting from internet: kubernetes pause image (tagged) + any images required for user workloads
```

WMCO will use the mirrors data in `ImageDigestMirrorSet` and `ImageTagMirrorSet` resources 
to create a set of [hosts.toml files](https://github.com/containerd/containerd/blob/main/docs/hosts.md#hoststoml-content-description---detail),
one for each registry mirror, that will be transferred to the Windows instance. WMCO will handle the file content 
creation and will copy over the files to each Windows node.

Since WMCO is an day 2 operator (optional, not part of the OCP payload), it will pick up the mirror configuration during
runtime regardless of whether they were set during cluster install time or at some point during the cluster's lifetime.

#### Containerd Config File Changes

WMCO will need to set the `config_path` variable in the containerd config file to tell it where to look for registry
config settings; it is currently empty in value. We should likely set it to `C:/k/containerd/registries` -- this is 
within our managed directories and ensures the directory contents are cleaned up during node deconfiguration.

#### Managing Containerd Registry Config Files

WMCO will generate the required registry config files based on the IDMS/ITMS resources in the cluster.

> For new nodes, these files will be copied over to the instances on configuration, similarly to how other config files

Reviewer 5: similarly-> similar

---

> title: winc-support-in-disconnected-environments

Reviewer: Should this be reframed as Windows node support for image mirroring?
All the work seems to be around this

Author: I was wrestling with this as well but was not 100% sure since we are adding support for specific image mirroring methods only, the minimum required to support disconnected envs. The Non-Goals section outlines this, let me know if you think we should reframe still

and users [should migrate to IDMSes](https://docs.openshift.com/container-platform/4.14/openshift_images/image-configuration.html#images-configuration-registry-mirror-convert_image-configuration) instead.
* Using [`ImageContentPolicy` resources](https://docs.openshift.com/container-platform/4.14/rest_api/config_apis/imagecontentpolicy-config-openshift-io-v1.html)

## Proposal

### Design Details

WMCO will use the mirrors data in `ImageDigestMirrorSet` and `ImageTagMirrorSet` resources 
to create a set of [hosts.toml files](https://github.com/containerd/containerd/blob/main/docs/hosts.md#hoststoml-content-description---detail),
one for each registry mirror, that will be transferred to the Windows instance.
This will be done on during initial node configuration, similarly to how bootstrap files (like the kubelet config file)
are generated and transferred. This way nodes will have the right mirror registry config right when they join the cluster.

In addition, the operator should also react to any events on IDMS/ITMS resources through a new "registry controller" --
when mirror settings are updated, WMCO will then react by:
- generating an updated set of hosts.toml files

> - copying over the new registry config files to each Windows instance, if a file has changed

Reviewer: This is a pattern we should be careful with. WMCO should avoid SSHing into Windows nodes to react to cluster changes. Think about this with a large amount of nodes, and frequent changes to the mirrors. Maintaining the state of dynamic configuration files is something we should make use of the windows-services configmap to do.

server = "https://docker.io"

[host."https://docker-mirror.internal"]
 capabilities = ["pull", "resolve"] # resolve tags
```

---

## Test Plan

In addition to unit testing individual WMCO/WICD packages and controllers, an e2e job will be added to the release repo
for WMCO's master/release-4.15 branches. A new CI workflow will be created leveraging 
[existing step-registry steps](https://github.com/openshift/release/tree/master/ci-operator/step-registry/aws/provision/vpc/disconnected),
which creates a disconnected cluster in AWS with hybrid-overlay networking. 
I propose AWS here instead of vSphere since we have 3+ existing jobs that run on vSphere and I don't want to overload the infra.
We'll also need to pre-load the images used in our e2e suite into the cluster's internal registry as part of job setup.
This workflow will be used to run the existing WMCO e2e test suite to ensure that WMCO runs as expected in disconnected 
environments with no regression. We will add a few test cases to explicitly check the state of registry settings on Windows
nodes.
QE should cover disconnected environments on all supported platforms when validating this feature.

### User Stories

> Story 1: Set `config_path` in containerd_conf.toml

Reviewer 2: The user stories should be of the form: `As an OpenShift Windows Admin, I should be able to define mirrors for my cluster, so my cluster would function when  disconnected from primary registries`.

Author: ACK, I've added in details in the proper form.

> Story 2: WMCO generates hosts.toml files

Reviewer 2: The stories as written should be moved to proposal/design/impl details

Story 3: WMCO copies over generated hosts.toml files to each Windows instance

User stories can also be found within the JIRA epic: [WINC-936](https://issues.redhat.com/browse/WINC-936)

### Workflow Description and Variations

**cluster creator** is a human user responsible for deploying a cluster.
**cluster administrator** is a human user responsible for managing cluster settings including network egress policies.

There are 4 different workflows that affect the disconnected use case.

> 1. A cluster creator brings up a new disconnected cluster and configures mirror registry settings at install time

Reviewer 2: These sound more like the user stories

In cases 1, 2 & 3, Windows nodes can be joined to the cluster after altering registry settings,
which would result in WMCO applying mirror settings during initial node configuration. 

In case 2, for any existing Windows nodes, WMCO will react by updating each instance's state to consume the new registry config.

In case 3, existing Windows nodes will not be re-configured. Existing registry settings will stay applied.

In case 4, WMCO will take no action as there are no settings to configure. Windows nodes will not be able to pull images
until mirror registry settings are configured by the administrator.

### Risks and Mitigations

If node cleanup fails, instances may retain their containerd configuration. This could be a concern for BYOH instances
particularly. An event should be generated to alert cluster admins manual cleanup is required to remove WMCO-created files.

Also, WMCO has no way of detecting local changes to the containerd registry config directory on Windows nodes.
The changes will remain in effect until a new IDMS/ITMS change is made in the cluster or the node is reconfigured.

> Another concern is if a user's mirror configuration has overlapping settings/clashes. As on the Linux side, it is upon

Reviewer 2: Not sure what we mean by this? Example?

Author: Added an example, hope it makes it more clear

the user to ensure there is no clash between mirrors across their IDMS and ITMS resources.

<!--
Containerd may not support all the fields available in IDMS/ITMS resources.
Unsupported fields will be ignored and will not impact containerd registry config/default behavior.
-->

### Drawbacks

> The major drawback with this design is a scaling issue. WMCO will have to SSH into each node to update mirror registry

Reviewer 2: Don't we have a per node entity that can watch and apply this concurrently?

Author: Yes we do, the Windows Instance Config Daemon (WICD) runs as a service on each Windows node. The previous draft of the enhancement actually followed this design
- WMCO consumes cluster mirror settings to generate the containerd config files
- WMCO publishes the files to WICD (through a configmap that WICD watches)
- WICD consumes configmap data to create files on Windows node, updating mirror settings

This would require us to create a mechanism similar to how MCO stores files in MachineConfigs. The team had a meeting around this and concluded (with Aravindh) that this is too complex of an approach for the disconnected MVP. We decided to implement the simpler solution, and then take on a future epic to introduce a `MachineConfigSpec`-like object.

TLDR: we heavily considered it but eventually decided against it

Reviewer 2: How many maximum number of nodes can we support with the current approach? That's something we would want documented.

Author: I don't think we can put a number on it, but the time it takes to update mirror settings will scale linearly with the # of Windows nodes in the cluster in the worst case. IMO this won't affect the # nodes we can support from a technical standpoint. The limit more so depends on a user's tolerance for longer re-configuration times and how often do they change their mirror settings.

One thing to note is that we are introducing a new controller that will only handle events related to image mirroring. This means that we won't have to enter into a full node re-configuration when cluster mirror settings change.

Reviewer 2: I think we can proceed with clear documentation covering this and then if this becomes an issue for adoption, we should get the other work prioritized to make it scale.

## Proposal

### Design Details

There are a couple pieces of work as part of this enhancement. All changes detailed in this enhancement proposal will be
limited to the Windows Machine Config Operator and its sub-component, Windows Instance Config Daemon (WICD).

First, WMCO will set the `config_path` variable in the containerd config file.
Next, the operator will generate registry config files based on cluster resources. 
And lastly, WMCO will place the generated files in the containerd config directory on each Windows instance.

WMCO will use the mirrors data in `ImageDigestMirrorSet` and `ImageTagMirrorSet` resources 
to create a set of [hosts.toml files](https://github.com/containerd/containerd/blob/main/docs/hosts.md#hoststoml-content-description---detail),
one for each registry mirror, that will be transferred to the Windows instance. WMCO will handle the file content 
creation and will copy over the files to each Windows node.

Since WMCO is an day 2 operator (optional, not part of the OCP payload), it will pick up the mirror configuration during
runtime regardless of whether they were set during cluster install time or at some point during the cluster's lifetime.

> #### Setting `config_path`

Reviewer: Containerd config file changes

WMCO will need to set the `config_path` variable in the containerd config file to tell it where to look for registry
config settings; it is currently empty in value. We should likely set it to `C:/k/containerd/registries` -- this is 
within our managed directories and ensures the directory contents are cleaned up during node deconfiguration.

> #### Config File Management

Reviewer: Management of Containerd Registry Config Files

WMCO will generate the required registry config files based on the IDMS/ITMS resources in the cluster.
For new nodes, these files will be copied over to the instances on configuration, similarly to how other config files
(kublet conf, bootstrap kubeconfig, files from ignition, etc) are treated today.

For existing nodes, the operator will react to updates to the cluster's mirror settings through a new controller
which will watch for events on IDMS/ITMS resources. This "registry controller" will generate and copy over the new
registry config files to each Windows instance when an event is detected.
We will fully replace the contents of the registry config directory on any change; this way, out of date registry config

> files will be removed without any extra mechanism to track files we have created in the past.

Reviewer: > on any change

On any change to the IDMS/ITMS cluster objects.

This distinction is important as if a local change to the directory is made, WMCO has no way of detecting this, and the change will remain until a new IDMS/ITMS change is made in the cluster.

Please cover this in the risks/drawbacks section.

Author: > This distinction is important as if a local change to the directory is made, WMCO has no way of detecting this, and the change will remain until a new IDMS/ITMS change is made in the cluster.

Added this to the Risks section

which would result in WMCO applying mirror settings during initial node configuration. 

In case 2, for any existing Windows nodes, WMCO will react by updating each instance's state to consume the new registry config.

In case 3, existing Windows nodes will not be re-configured. Existing registry settings will stay applied.

In case 4, WMCO will take no action as there are no settings to configure. Windows nodes will not be able to pull images
until mirror registry settings are configured by the administrator.

### Risks and Mitigations

If node cleanup fails, instances may retain their containerd configuration. This could be a concern for BYOH instances
particularly. An event should be generated to alert cluster admins manual cleanup is required to remove WMCO-created files.

Another concern is if a user's mirror configuration has overlapping settings/clashes. As on the Linux side, it is upon
the user to ensure there is no clash between mirrors across their IDMS and ITMS resources.

<!--
Containerd may not support all the fields available in IDMS/ITMS resources.
Unsupported fields will be ignored and will not impact containerd registry config/default behavior.
-->

### Drawbacks

> The only drawbacks are the increased complexity of WMCO and the potential complexity of debugging customer cases that

Reviewer: You need to call out the scaling issues due to the sequential nature of copying the files on IDMS/ITMS change.

---
title: winc-support-in-disconnected-environments
authors:
  - "@saifshaikh48"
reviewers:
  - "@openshift/openshift-team-windows-containers"
  - "@openshift/openshift-team-pod-autoscaling, for insight on MCO's registry configuration on Linux nodes"
approvers:

>   - "@aravindhp"

Reviewer 6: should Aravindh be an approver, or should Sebastian?

* Supporting clusters that use only `ImageContentSourcePolicy` resources. These were [deprecated in 4.13](https://docs.openshift.com/container-platform/4.14/release_notes/ocp-4-14-release-notes.html#ocp-4-14-deprecated-removed-features) 
and users [should migrate to IDMSes](https://docs.openshift.com/container-platform/4.14/openshift_images/image-configuration.html#images-configuration-registry-mirror-convert_image-configuration) instead.
* Using [`ImageContentPolicy` resources](https://docs.openshift.com/container-platform/4.14/rest_api/config_apis/imagecontentpolicy-config-openshift-io-v1.html)

## Proposal

### Design Details

WMCO will use the mirrors data in `ImageDigestMirrorSet` and `ImageTagMirrorSet` resources 
to create a set of [hosts.toml files](https://github.com/containerd/containerd/blob/main/docs/hosts.md#hoststoml-content-description---detail),
one for each registry mirror, that will be transferred to the Windows instance.

This will be done on during initial node configuration, similarly to how bootstrap files (like the kubelet config file)

> are generated. This way nodes will have the right mirror registry config right when they join the cluster.

Reviewer: Why does this need to be done?

Author: You're right I don't think it needs to be done. When the controllers start, things should kickstart properly on their own

In addition, the operator should also react to any events on IDMS/ITMS resources through a new "registry controller" --
when mirror settings are updated, WMCO will then react by:
- generating an updated set of hosts.toml files

> - publishing them to the windows-services ConfigMap as `Files` 

Reviewer: Please add more detail about the publishing. The windows-services configmap is `immutable`

The windows-services ConfigMap spec will have to be extended:
* we will add a `content` field the `files` key so it will look like
```json
{
  "path": "filepath",
  "checksum": "checksum of file, generated at compile time",
  "content": "base64 encoded file data, similarly to how MCO stores files in MachineConfigs",
}
```

WICD, which already watches the services ConfigMap, will react to events on it by copying over the new registry config
files to each Windows instance, if a file has changed.
Note that if the newly generated set of files is empty, this implies the cluster mirror regsitry settings have been cleared.

> WMCO will respond by deleting all directories and hosts.toml files under containerd's `config_path` directory.

Reviewer: This sounds like its special casing the config_path directory.
If that is the case, please avoid this by altering the windows-services configmap schema to allow a way to express what you need to do.

Author: Introduced "FilesToDelete" to address this

We can utilize [registry runtime-utils](https://github.com/openshift/runtime-utils/blob/master/pkg/registries/registries.go)
to easily process and merge the contents of the IDMS/ITMS resources, [as MCO does](https://github.com/openshift/machine-config-operator/blob/master/pkg/controller/container-runtime-config/helpers.go#L407).

The directory to save the generated registry config files is defined by the containerd_config.toml's
`plugins."io.containerd.grpc.v1.cri".registry.config_path`.

The containerd daemon [picks up the registry configuration without requiring restart](https://github.com/containerd/containerd/blob/main/docs/hosts.md#registry-configuration---introduction).

Since WMCO is an day 2 operator (optional, not part of the OCP payload), it will pick up the mirror configuration during

> runtime regardless of whether they were set during cluster install time or at some point during the cluster's lifetime.

Reviewer: This section feels disjointed, and like an information dump. I suggest adding subheaders to help compartmentalize the info you are giving.

Author: On second read, I totally agree. I reworked it and broke it up into separate sections

All changes detailed in this enhancement proposal will be limited to the Windows Machine Config Operator and its 

> sub-component, Windows Instance Config Daemon (WICD).

Reviewer: Shouldnt this be at the beginning of the section?

> Example: A cluster contains the following resources

Reviewer: Please move this into a subsection.

There already exists a protocol for users to publish [registry mirroring configuration](https://docs.openshift.com/container-platform/4.14/openshift_images/image-configuration.html#images-configuration-registry-mirror_image-configuration),
namely `ImageDigestMirrorSet` (IDMS), `ImageTagMirrorSet` (ITMS) cluster resources. 
These are consumed by OpenShift components like the Machine Config Operator (MCO) to apply the settings to Linux
control-plane and worker nodes, Windows worker nodes do not currently consume or respect mirror registry settings when
pulling images. This effort will work to plug feature disparity by making the WMCO aware of mirror registry settings at
operator install time and reactive during its runtime.

## Motivation

The motivation here is to expand the Windows containers production use case, enabling users to add Windows nodes and run
workloads easily and successfully in disconnected clusters. This is an extremely important ask for customer environments
where Windows nodes must pull images from air-gapped registries for security reasons.
The added benefit here is that the changes proposed in this enhancement are not restricted to disconnected environments;

> users in __connected__ clusters can take advantage of this mirror registry support. This would allow better performance

Reviewer 3: Does mirror registry support imply that customers can configure Windows nodes to pull from other image registries that don't contain the mirror of all the images required for disconnected support?

Author: yes, customers would be able to configure other image registries as mirrors as well

> as images can be stored in registries closer to the production environment, reducing pull times for notoriously large 

Reviewer 3: Please remove `notoriously`. There is no reason for such adjectives.

Windows container images, in addition the security benefits of restricted networks.

### Goals

* Create a mechanism for WMCO to consume global mirror registry settings from existing platform resources (IDMS/ITMSes),
  such that it enables Windows nodes to pull images in disconnected environments

> * Configure the mirror registry settings in the containerd runtime, a WMCO-managed component on Windows nodes 

Reviewer 3: Is this mirror registry settings or just registry settings from a containerd perspective?

Author: from containerd's perspective these are just registry settings, but WMCO will only be setting those registry settings related to mirrors

* React to changes to the mirror registry settings during WMCO runtime
* Maintain normal functionality in connected clusters and clusters without any cluster-wide mirror registry configuration

### Non-Goals

* First-class support of [all OpenShift image registry setting configurations](https://docs.openshift.com/container-platform/4.14/openshift_images/image-configuration.html#images-configuration-file_image-configuration), including the `Image` resource
* Supporting clusters that use only `ImageContentSourcePolicy` resources. These were [deprecated in 4.13](https://docs.openshift.com/container-platform/4.14/release_notes/ocp-4-14-release-notes.html#ocp-4-14-deprecated-removed-features) 
and users [should migrate to IDMSes](https://docs.openshift.com/container-platform/4.14/openshift_images/image-configuration.html#images-configuration-registry-mirror-convert_image-configuration) instead.
* Using [`ImageContentPolicy` resources](https://docs.openshift.com/container-platform/4.14/rest_api/config_apis/imagecontentpolicy-config-openshift-io-v1.html)

## Proposal

### Design Details

> There are a couple pieces of work as part of this enhacement. All changes detailed in this enhancement proposal will be

Reviewer 5: enhancement*

limited to the Windows Machine Config Operator and its sub-component, Windows Instance Config Daemon (WICD).
First, WMCO will set the `config_path` variable in the containerd config file.
Next, the operator will generate registry config files based on cluster resources and publish them to WICD. 

> And lastly, WICD will consume the generate registry config and create them on each Windows instance.

Reviewer 5: generate-> generated

WMCO will use the mirrors data in `ImageDigestMirrorSet` and `ImageTagMirrorSet` resources 
to create a set of [hosts.toml files](https://github.com/containerd/containerd/blob/main/docs/hosts.md#hoststoml-content-description---detail),
one for each registry mirror, that will be transferred to the Windows instance. WMCO will handle the file content 

> creation and will publish the required info to WICD. WICD will consume the given info and create/update files on the

Reviewer 3: By publish do you mean add it to the services ConfigMap?

Windows instances.

Since WMCO is an day 2 operator (optional, not part of the OCP payload), it will pick up the mirror configuration during
runtime regardless of whether they were set during cluster install time or at some point during the cluster's lifetime.

#### Setting `config_path`

WMCO will need to set the `config_path` variable in the containerd config file to tell it where to look for registry
config settings; it is currently empty in value. We should likely set it to something within our managed directories

> (e.g. "C:\\k") to ensure they're cleaned up during node deconfiguration.

Reviewer 3: > We should likely set it to something within our managed directorie

This is the design where you decide what you are going to do. Please be specific.

#### File Content

WMCO will generate the required registry config files based on the IDMS/ITMS resources in the cluster.
The operator will react to updates to the cluster's mirror settings through a new "registry controller" which will
watch for events on IDMS/ITMS resources.
WMCO will publish the generated file content into the windows-services ConfigMap.

> Because the windows-services ConfigMap is immutable, publishing changes to it will require a couple steps.

Reviewer 3: ```suggestion
Because the windows-services ConfigMap is immutable, publishing changes to it will require the following steps:
```

> The file content management flow will be as such:

Reviewer 3: Please add links to the code if you are getting down to that level detail. I feel you should focus on it from a higher level rather than go down to the implementation details here.

Reviewer 3: Have you taken into account the operator crashing in the middle of any of these operations? Will it be able to recover?

1. First, we'll introduce a method to update the configmap_controller's `servicesManifest` object which holds the
  windows-services CM data
2. then, generate an updated set of hosts.toml files
3. next, compute a diff between the `Files` in the existing windows-services ConfigMap and the newly generated files
  - any files in the existing spec that are NOT in the new set will will be added to the serviceManifest's `FilesToDelete`
  using the method introduced in step 1
4. update the serviceManifest's `Files` to hold the newly generated set of hosts.toml files
5. lastly, publish the new windows-services ConfigMap
  - Delete the cluster's existing windows-services ConfigMap from the registry controller
  - The configmap_controller will react by re-creating the windows-services ConfigMap (as it does today), with new spec

> The windows-services ConfigMap spec will have to be extended. We will add a mandatory `content` field to the `files` key

Reviewer 3: Is this done for the CNI config? If not, why is this needed for the containderd / hosts.toml files?

as well as a new field `FilesToDelete` which is just a list of file paths:
```yaml
{
  Services: []Service{...},
  Files: []FileInfo{
    "path": "filepath",
    "checksum": "checksum of file, generated at compile time",
    "content": "base64 encoded file data, similarly to how MCO stores files in MachineConfigs",
  },
  FilesToDelete: []string{...} ,
  EnvironmentVars: map[string]string{...},
  WatchedEnvironmentVars: []string{...},
}
```

We can utilize [registry runtime-utils](https://github.com/openshift/runtime-utils/blob/master/pkg/registries/registries.go)
to easily process and merge the contents of the IDMS/ITMS resources, [as MCO does](https://github.com/openshift/machine-config-operator/blob/master/pkg/controller/container-runtime-config/helpers.go#L407).

> For registry config, I don't forsee any of the files having sensitive information. If new information arises suggesting

Reviewer 3: What if the registry config require authentication?

otherwise, we can leverage our `crypto.go` library to safely encrypt the file content before placing it in the
windows-services ConfigMap.

The directory to save the generated registry config files is defined by WMCO -- it should be the same as what it set for
the containerd_conf.toml's `plugins."io.containerd.grpc.v1.cri".registry.config_path`. 

#### File Creation

> WICD, which already watches the windows-services ConfigMap, will react to events on it by copying over the newly

Reviewer 3: How will WICD copy the file to each Windows instance?

generated registry config files to each Windows instance, if a file has changed.

The containerd daemon [picks up the registry configuration without requiring restart](https://github.com/containerd/containerd/blob/main/docs/hosts.md#registry-configuration---introduction) so no further action required.

#### File Cleanup

> WICD will look at the `FilesToDelete` field of the windows-services ConfigMap and ensure files at those locations are deleted.

Reviewer 5: what happens if the cleanup fails?

Mirror registry directory and files will be deleted on node cleanup as part of WMCO's `RemoveFilesAndNetworks()` function.

#### Working example

Example: A cluster contains the following resources
```yaml
apiVersion: config.openshift.io/v1 
kind: ImageDigestMirrorSet 
metadata:
  name: ubi9repo
spec:
  imageDigestMirrors: 
  - mirrors:
    - example.io/example/ubi-minimal
    - example.com/example2/ubi-minimal
    source: registry.access.redhat.com/ubi9/ubi-minimal 
    mirrorSourcePolicy: AllowContactingSource 
  - mirrors:
    - mirror.example.com
    source: registry.redhat.io
    mirrorSourcePolicy: NeverContactSource
```

```yaml
apiVersion: config.openshift.io/v1
kind: ImageTagMirrorSet
metadata:
  name: tag-mirror
spec:
  imageTagMirrors:
  - mirrors:
    - docker.io
    source: docker-mirror.internal
    mirrorSourcePolicy: AllowContactingSource
```

would result in WMCO generating content for 3 files to transfer onto each Windows instance. The new windows-services
ConfigMap that WMCO creates will have:

```yaml
{
  Services: []Service{...},

>   Files: []FileInfo{

Reviewer 3: Even after looking at the example I am not clear why this toml file is being treated differently when compared to the other config files for the node.

In addition to unit testing individual WMCO/WICD packages and controllers, an e2e job will be added to the release repo
for WMCO's master/release-4.15 branches. A new CI workflow will be created leveraging 
[existing step-registry steps](https://github.com/openshift/release/tree/master/ci-operator/step-registry/aws/provision/vpc/disconnected),
which creates a disconnected cluster in AWS with hybrid-overlay networking. 
I propose AWS here instead of vSphere since we have 3+ existing jobs that run on vSphere and I don't want to overload the infra.
We'll also need to pre-load the images used in our e2e suite into the cluster's internal registry as part of job setup.
This workflow will be used to run the existing WMCO e2e test suite to ensure that WMCO runs as expected in disconnected 
environments with no regression. We will add a few test cases to explicitly check the state of registry settings on Windows
nodes.
QE should cover disconnected environments on all supported platforms when validating this feature.

### User Stories

Story 1: Set `config_path` in containerd_conf.toml
Story 2: WMCO generates hosts.toml files and adds them into services CM spec

> Story 3: WICD copies over/reconciles files from services CM

Reviewer 3: These are not user stories perse and I suggest just pointing at the epic.

involve a custom mirror registry setup, since it would be harder to set up an accurate replication environment.
This can be mitigated by proactively getting the development team, QE, and support folks familiar with the expected
behavior of Windows nodes/workloads in disconnected clusters.

#### Support Procedures

In general the support procedures for WMCO will remain the same.
Customers that need support with disconnected environments must upgrade to the minimum versions: OCP 4.15 and WMCO 10.15.0.

### API Extensions

N/A, as no CRDs, admission and conversion webhooks, aggregated API servers, or finalizers will be added or modified.
Only the WMCO will be extended which is an optional operator with its own lifecycle and SLO/SLAs, a
[tier 3 OpenShift API](https://docs.openshift.com/container-platform/4.12/rest_api/understanding-api-support-tiers.html#api-tiers_understanding-api-tiers).

### Operational Aspects of API Extensions

N/A

#### Failure Modes

N/A

### Release Plan

The feature associated with this enhancement is targeted to land in the official Red Hat operator version of WMCO 10.15.0

> within OpenShift 4.15 timeframe. The normal WMCO release process will be followed as the functionality described in this

Reviewer 5: Shouldn't this be 10.16/OpenShift 4.16? This changes the upgrade and version strategy below.

enhancement is integrated into the product.

> An Openshift docs update announcing Windows disconnected support will be required as part of GA. The new docs

Reviewer 3: We need a blog for this type of large user facing feature.

The official Red Hat operator in OCP 4.15, WMCO version 10.15.0, will support Windows containers in disconnected
environments. Releases 9.y.z and below will not have support disconnected support.

## Implementation History

The implementation history can be tracked by following the associated work items in Jira and source code improvements in
the WMCO Github repo.

## Alternatives & Justification
 
For disconnected support, we could make the kubelet pause image location configurable. This is a nice to have, but may 
be out of scope of disconnected support as it could require a larger undertaking to make the containerd config file
configurable.

There are a few alternatives that are no longer viable.
Pre-pulling images, where user creates a Windows image with the pause image and other required images baked in, is not
possible anymore as we install and manage the container runtime.
Also, users cannot configure containerd's registries manually since WMCO will overwrite changes to the containerd
config file if there is some unrelated event that triggers node reconciliation.

## Open Questions

> * Would it work "out-of-the-box" today if images were pre-loaded with specific tags into the OpenShift internal registry,

Reviewer 3: I don't understand this open question or it's implications

Author: It served mostly as an info dump of a workaround to get WMCO working in a disconnected env that required the least possible dev work. I will remove this

Minimal work required for us:
We'd have to set a `config_path` in the base containerd_conf.toml file to a valid value, as recommended in their docs.
It is currently set to `""` -- containerd log output:
```
Registry:{ConfigPath: Mirrors:map[] Configs:map[] Auths:map[] Headers:map[]}
```
This serves as the location of the _default/hosts.toml file, which, if the user set manually today, 
would be stomped back by WMCO if there were any unrelated events that triggers node reconciliation.

Bit of manual work on the user:
1. Pull and tag images in private registry (air-gapped, on-prem or openshift internal registry) before disconnecting from internet: kubernetes pause image (tagged) + any images required for user workloads
* ctr pull mcr.microsoft.com/oss/kubernetes/pause:3.9
* ctr --namespace=openshift image tag mcr.microsoft.com/oss/kubernetes/pause:3.9 image-registry.openshift-image-registry.svc:5000/openshift/pause:3.9
* ctr push image-registry.openshift-image-registry.svc:5000/openshift/pause:3.9

2. Create a `config_path`/_default/hosts.toml file. Example content:
```
[host."https://image-registry.openshift-image-registry.svc:5000"]
  capabilities = ["pull", "resolve"]

> ```

Reviewer 3: I don't follow why this is an open question
